src/web/demo/src/main/java/org/geoserver/web/demo/KMLFormatLink.java

```diff
@@ -14,7 +14,7 @@ public ExternalLink getFormatLink(PreviewLayer layer) {
         ExternalLink kmlLink =
                 new ExternalLink(
                         this.getComponentId(),
-                        layer.getWmsLink() + "/kml?layers=" + layer.getName(),
```

Comment:
I think KML was supposed to use the reflector, the refactor that introduced the KMLFormatLink broke it.
Here is the link as it was supposed to be:
https://github.com/geoserver/geoserver/blob/2.15.x/src/web/demo/src/main/java/org/geoserver/web/demo/MapPreviewPage.java#L102

Looks like PreviewLayer still provides the methods needed there.
And then MapPreviewPageTest needs to be extended with a check on the links content (should be easy enough, try using ``print(tester.getLastRendereredPage(), true, true)`` to find out the wicket path of the component holding the link

Reply:
Oh, actually not sure that approach would work 100% well with URL manglers being around, that might alter the query string (e.g. authkey).
Probbaly best to get the base url from the preview layer (available) and then call the response builder with "layers" and the layer name as a KVP. Something similar to:

```java
ResponseUtils.buildURL(baseURL, service, Collections.singletonMap("layers", layer.getname), URLType.SERVICE)
```

Reply:
I don't see that use of the reflector is important, and haven't incorporated that.

I have refactored the URL construction, and also applied the same logic to OpenLayers. Manual testing shows it is fine for both, and there are some unit test adds / updates to match.

Reply:
> I don't see that use of the reflector is important, and haven't incorporated that.

It's actually pretty different from a WMS call. A plain WMS call sets the KML generation to "download" mode. The "poi"  example layer in GeoServer generates a fully self contained KML with all the data:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<kml xmlns="http://www.opengis.net/kml/2.2" xmlns:ns2="http://www.google.com/kml/ext/2.2" xmlns:ns3="http://www.w3.org/2005/Atom" xmlns:ns4="urn:oasis:names:tc:ciq:xsdschema:xAL:2.0">
    <Document>
        <LookAt>
            <longitude>-74.00668149807987</longitude>
            <latitude>40.713716095854025</latitude>
            <altitude>2009.2659535747807</altitude>
            <heading>0.0</heading>
            <tilt>0.0</tilt>
            <range>1623.592279843792</range>
            <altitudeMode>clampToGround</altitudeMode>
        </LookAt>
        <Folder>
            <name>tiger:poi</name>
            <LookAt>
                <longitude>-74.00668149807987</longitude>
                <latitude>40.713716095854025</latitude>
                <altitude>2009.2659535747807</altitude>
                <heading>0.0</heading>
                <tilt>0.0</tilt>
                <range>1623.592279843792</range>
                <altitudeMode>clampToGround</altitudeMode>
            </LookAt>
            <Placemark id="poi.1">
                <name>museam</name>
                <description>&lt;h4&gt;poi&lt;/h4&gt;

&lt;ul class="textattributes"&gt;
  
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;NAME&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;museam&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;THUMBNAIL&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037827-Ti.jpg&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;MAINPAGE&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037827-L.jpg&lt;/span&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
                <LookAt>
                    <longitude>-74.01046109936334</longitude>
                    <latitude>40.70758762625656</latitude>
                    <altitude>0.0</altitude>
                    <heading>0.0</heading>
                    <tilt>0.0</tilt>
                    <range>0.0</range>
                    <altitudeMode>clampToGround</altitudeMode>
                </LookAt>
                <Style>
                    <IconStyle>
                        <scale>0.6875</scale>
                        <heading>0.0</heading>
                        <Icon>
                            <href>http://localhost:8080/geoserver/kml/icon/poi?0.0.0=&amp;0.0.1=</href>
                        </Icon>
                    </IconStyle>
                    <LabelStyle>
                        <color>ff000000</color>
                        <scale>1.4</scale>
                    </LabelStyle>
                </Style>
                <Point>
                    <coordinates>-74.01046109936333,40.707587626256554</coordinates>
                </Point>
            </Placemark>
            <Placemark id="poi.2">
                <name>stock</name>
                <description>&lt;h4&gt;poi&lt;/h4&gt;

&lt;ul class="textattributes"&gt;
  
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;NAME&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;stock&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;THUMBNAIL&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037829-Ti.jpg&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;MAINPAGE&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037829-L.jpg&lt;/span&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
                <LookAt>
                    <longitude>-74.0108375113659</longitude>
                    <latitude>40.70754683896325</latitude>
                    <altitude>0.0</altitude>
                    <heading>0.0</heading>
                    <tilt>0.0</tilt>
                    <range>0.0</range>
                    <altitudeMode>clampToGround</altitudeMode>
                </LookAt>
                <Style>
                    <IconStyle>
                        <scale>0.6875</scale>
                        <heading>0.0</heading>
                        <Icon>
                            <href>http://localhost:8080/geoserver/kml/icon/poi?0.0.0=&amp;0.0.1=</href>
                        </Icon>
                    </IconStyle>
                    <LabelStyle>
                        <color>ff000000</color>
                        <scale>1.4</scale>
                    </LabelStyle>
                </Style>
                <Point>
                    <coordinates>-74.0108375113659,40.70754683896324</coordinates>
                </Point>
            </Placemark>
            <Placemark id="poi.3">
                <name>art</name>
                <description>&lt;h4&gt;poi&lt;/h4&gt;

&lt;ul class="textattributes"&gt;
  
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;NAME&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;art&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;THUMBNAIL&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037856-Ti.jpg&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;MAINPAGE&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037856-L.jpg&lt;/span&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
                <LookAt>
                    <longitude>-74.01053023879956</longitude>
                    <latitude>40.70938711687079</latitude>
                    <altitude>0.0</altitude>
                    <heading>0.0</heading>
                    <tilt>0.0</tilt>
                    <range>0.0</range>
                    <altitudeMode>clampToGround</altitudeMode>
                </LookAt>
                <Style>
                    <IconStyle>
                        <scale>0.6875</scale>
                        <heading>0.0</heading>
                        <Icon>
                            <href>http://localhost:8080/geoserver/kml/icon/poi?0.0.0=&amp;0.0.1=</href>
                        </Icon>
                    </IconStyle>
                    <LabelStyle>
                        <color>ff000000</color>
                        <scale>1.4</scale>
                    </LabelStyle>
                </Style>
                <Point>
                    <coordinates>-74.01053023879955,40.70938711687079</coordinates>
                </Point>
            </Placemark>
            <Placemark id="poi.4">
                <name>lox</name>
                <description>&lt;h4&gt;poi&lt;/h4&gt;

&lt;ul class="textattributes"&gt;
  
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;NAME&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;lox&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;THUMBNAIL&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037884-Ti.jpg&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;MAINPAGE&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037884-L.jpg&lt;/span&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
                <LookAt>
                    <longitude>-74.00857344353275</longitude>
                    <latitude>40.711945649065406</latitude>
                    <altitude>0.0</altitude>
                    <heading>0.0</heading>
                    <tilt>0.0</tilt>
                    <range>0.0</range>
                    <altitudeMode>clampToGround</altitudeMode>
                </LookAt>
                <Style>
                    <IconStyle>
                        <scale>0.6875</scale>
                        <heading>0.0</heading>
                        <Icon>
                            <href>http://localhost:8080/geoserver/kml/icon/poi?0.0.0=&amp;0.0.1=</href>
                        </Icon>
                    </IconStyle>
                    <LabelStyle>
                        <color>ff000000</color>
                        <scale>1.4</scale>
                    </LabelStyle>
                </Style>
                <Point>
                    <coordinates>-74.00857344353275,40.711945649065406</coordinates>
                </Point>
            </Placemark>
            <Placemark id="poi.5">
                <name>church</name>
                <description>&lt;h4&gt;poi&lt;/h4&gt;

&lt;ul class="textattributes"&gt;
  
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;NAME&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;church&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;THUMBNAIL&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037839-Ti.jpg&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;MAINPAGE&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/22037839-L.jpg&lt;/span&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
                <LookAt>
                    <longitude>-74.0118315772888</longitude>
                    <latitude>40.708529961953786</latitude>
                    <altitude>0.0</altitude>
                    <heading>0.0</heading>
                    <tilt>0.0</tilt>
                    <range>0.0</range>
                    <altitudeMode>clampToGround</altitudeMode>
                </LookAt>
                <Style>
                    <IconStyle>
                        <scale>0.6875</scale>
                        <heading>0.0</heading>
                        <Icon>
                            <href>http://localhost:8080/geoserver/kml/icon/poi?0.0.0=&amp;0.0.1=</href>
                        </Icon>
                    </IconStyle>
                    <LabelStyle>
                        <color>ff000000</color>
                        <scale>1.4</scale>
                    </LabelStyle>
                </Style>
                <Point>
                    <coordinates>-74.0118315772888,40.708529961953786</coordinates>
                </Point>
            </Placemark>
            <Placemark id="poi.6">
                <name>fire</name>
                <description>&lt;h4&gt;poi&lt;/h4&gt;

&lt;ul class="textattributes"&gt;
  
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;NAME&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;fire&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;THUMBNAIL&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/28640984-Ti.jpg&lt;/span&gt;&lt;/li&gt;
  &lt;li&gt;&lt;strong&gt;&lt;span class="atr-name"&gt;MAINPAGE&lt;/span&gt;:&lt;/strong&gt; &lt;span class="atr-value"&gt;pics/28640984-L.jpg&lt;/span&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
                <LookAt>
                    <longitude>-74.00153046439814</longitude>
                    <latitude>40.719885123828675</latitude>
                    <altitude>0.0</altitude>
                    <heading>0.0</heading>
                    <tilt>0.0</tilt>
                    <range>0.0</range>
                    <altitudeMode>clampToGround</altitudeMode>
                </LookAt>
                <Style>
                    <IconStyle>
                        <scale>0.6875</scale>
                        <heading>0.0</heading>
                        <Icon>
                            <href>http://localhost:8080/geoserver/kml/icon/poi?0.0.0=&amp;0.0.1=</href>
                        </Icon>
                    </IconStyle>
                    <LabelStyle>
                        <color>ff000000</color>
                        <scale>1.4</scale>
                    </LabelStyle>
                </Style>
                <Point>
                    <coordinates>-74.00153046439813,40.719885123828675</coordinates>
                </Point>
            </Placemark>
        </Folder>
    </Document>
</kml>
```

This works fine for small layers, but it's unsuitable for anything larger. Besides the large generation time, it used to crash Google Earth back at the times, not sure what happends today.
If you follow the kml reflector, e.g. ``http://localhost:8080/geoserver/wms/kml?layers=tiger:poi``, then the response is in "refresh" mode, looking like this:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<kml xmlns="http://www.opengis.net/kml/2.2" xmlns:ns2="http://www.google.com/kml/ext/2.2" xmlns:ns3="http://www.w3.org/2005/Atom" xmlns:ns4="urn:oasis:names:tc:ciq:xsdschema:xAL:2.0">
    <Document>
        <LookAt>
            <longitude>-74.01020256421765</longitude>
            <latitude>40.70974625546435</latitude>
            <altitude>694.1959814771428</altitude>
            <heading>0.0</heading>
            <tilt>0.0</tilt>
            <range>560.9467647722848</range>
            <altitudeMode>clampToGround</altitudeMode>
        </LookAt>
        <NetworkLink>
            <name>Manhattan (NY) points of interest</name>
            <visibility>1</visibility>
            <open>1</open>
            <description>Points of interest in New York, New York (on Manhattan). One of the attributes contains the name of a file with a picture of the point of interest.</description>
            <LookAt>
                <longitude>-74.00668161555254</longitude>
                <latitude>40.71371615223534</latitude>
                <altitude>2173.081355567034</altitude>
                <heading>0.0</heading>
                <tilt>0.0</tilt>
                <range>1755.96371704499</range>
                <altitudeMode>clampToGround</altitudeMode>
            </LookAt>
            <Url>
                <href>http://localhost:8080/geoserver/wms?service=wms&amp;request=GetMap&amp;version=1.1.1&amp;format=application/vnd.google-earth.kml+xml&amp;layers=tiger:poi&amp;styles=poi&amp;height=2048&amp;width=2048&amp;transparent=false&amp;srs=EPSG:4326&amp;format_options=AUTOFIT:true;KMATTR:true;KMPLACEMARK:false;KMSCORE:40;MODE:refresh;SUPEROVERLAY:false</href>
                <refreshInterval>0.0</refreshInterval>
                <viewRefreshMode>onStop</viewRefreshMode>
                <viewRefreshTime>1.0</viewRefreshTime>
                <viewBoundScale>1.0</viewBoundScale>
            </Url>
        </NetworkLink>
    </Document>
</kml>
```

The backlink is dynamic, GE receives new content based on the current view, it will be raster if zoomed out and there is too much data, or full vector if the area is smaller and not likely to cause issues. It's also using the KML parameter defaults, you can see them injected in the backlink url.

More about GE refresh mode here:
* https://docs.geoserver.org/stable/en/user/services/wms/googleearth/features/kmlreflector.html#reflector-modes
* https://geoserver.geo-solutions.it/educational/en/google_earth/refresh.html

Reply:
Will rework it.